### PR TITLE
ARROW-3646: [Python] High-level IO API

### PR DESCRIPTION
**Pull request by pitrou**
Add the ``pa.input_stream`` and ``pa.output_stream`` factory functions.

### CODE REVIEW

File: python/doc/source/memory.rst [hunk 1]
```diff
@@ -70,11 +74,42 @@ required, and such conversions are also zero-copy:
...
+All memory allocations and deallocations (like ``malloc`` and ``free`` in C)
+are tracked in an instance of ``arrow::MemoryPool``. This means that we can
+then precisely track amount of memory that has been allocated:
+
+.. ipython:: python
+
+   pa.total_allocated_bytes()
+
+PyArrow uses a default built-in memory pool, but in the future there may be
+additional memory pools (and subpools) to choose from.  Let's allocate
```
Comment (kszucs) [inline python/doc/source/memory.rst]:
```suggestion
additional memory pools (and subpools) to choose from. Let's allocate
```

Comment (pitrou) [reply python/doc/source/memory.rst]: Putting two spaces after a period is a [common convention](https://devguide.python.org/documenting/#use-of-whitespace). I'm not sure whether we have any guidelines on that. @wesm @xhochy

File: python/pyarrow/tests/test_io.py [hunk 2]
```diff
@@ -1050,3 +1055,218 @@ def test_compressed_roundtrip(compression):
...
+    data = b"some test data\n" * 10 + b"eof\n"
+    file_path = tmpdir / 'input_stream'
+    with open(str(file_path), 'wb') as f:
+        f.write(data)
+
+    stream = pa.input_stream(file_path)
+    assert stream.read() == data
+    stream = pa.input_stream(str(file_path))
+    assert stream.read() == data
+    if pathlib is not None:
```
Comment (kszucs) [inline python/pyarrow/tests/test_io.py]: Pathlib is already present in the testing suite https://github.com/apache/arrow/blob/e26806eeb4dcaa07b2f7dd7a7b3da2d0fdb42276/python/pyarrow/tests/conftest.py#L20-L23

File: python/pyarrow/tests/test_io.py [hunk 3]
```diff
@@ -1050,3 +1055,218 @@ def test_compressed_roundtrip(compression):
...
+        assert stream.read() == data
+
+    gz_data = gzip_compress(data)
+    stream = pa.input_stream(memoryview(gz_data))
+    assert stream.read() == gz_data
+    stream = pa.input_stream(memoryview(gz_data), compression='gzip')
+    assert stream.read() == data
+
+
+def test_input_stream_file_path(tmpdir):
```
Comment (kszucs) [inline python/pyarrow/tests/test_io.py]: Theres is a fixture called `tempdir` which is a `pathlib.Path` variant of pytest's `tmpdir` fixture.

File: python/pyarrow/tests/test_io.py (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/tests/test_io.py]: Ok, I see.

File: python/pyarrow/tests/test_io.py (hunk 3 above)
Comment (pitrou) [reply python/pyarrow/tests/test_io.py]: Yuck...

File: python/pyarrow/io.pxi [hunk 4]
```diff
@@ -1363,3 +1393,117 @@ def decompress(object buf, decompressed_size=None, codec='lz4',
...
+    elif isinstance(source, (Buffer, memoryview)):
+        stream = FixedSizeBufferWriter(as_buffer(source))
+    elif isinstance(source, BufferedIOBase):
+        if not source.writable():
+            raise TypeError("pa.output_stream() called with non-writable file")
+        stream = PythonFile(source, 'w')
+    elif file_type is not None and isinstance(source, file_type):
+        # Python 2 file type
+        if 'w' not in source.mode and '+' not in source.mode:
+            raise TypeError("pa.output_stream() called with non-writable file")
```
Comment (kszucs) [inline python/pyarrow/io.pxi]: We should move this validation to `PythonFile`

File: python/doc/source/memory.rst (hunk 1 above)
Comment (kszucs) [reply python/doc/source/memory.rst]: It seems like both conventions are used in the rst docs.

File: python/pyarrow/tests/test_io.py [hunk 5]
```diff
@@ -1050,3 +1055,218 @@ def test_compressed_roundtrip(compression):
...
+    assert check_data(file_path, data, compression=None) == data
+
+
+def test_output_stream_python_file(tmpdir):
+    data = b"some test data\n" * 10 + b"eof\n"
+
+    def check_data(data, **kwargs):
+        # XXX cannot use BytesIO because stream.close() is necessary
+        # to finish writing compressed data, but it will also close the
+        # underlying BytesiO
```
Comment (kszucs) [inline python/pyarrow/tests/test_io.py]: Typo `BytesIO`

File: python/pyarrow/io.pxi [hunk 6]
```diff
@@ -1363,3 +1393,117 @@ def decompress(object buf, decompressed_size=None, codec='lz4',
...
+    except TypeError:
+        source_path = None
+
+    if compression == 'detect':
+        if source_path is not None:
+            compression_type = _get_compression_type_by_filename(source_path)
+        else:
+            compression_type = CompressionType_UNCOMPRESSED
+    else:
+        compression_type = _get_compression_type(compression)
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/io.pxi]: Stream instantiation is slightly different. We could still factor output compression detection.

File: python/doc/source/memory.rst (hunk 1 above)
Comment (wesm) [reply python/doc/source/memory.rst]: AP Stylebook for journalism today is 1 space. 2 spaces seems to be a remnant of the typewriter era

**Review by kszucs — APPROVED** [review]
LGTM, merging on green

**Comment by pitrou** [comment]
Thanks for the review @kszucs !

**Comment by wesm** [comment]
I will have a look and leave my comments also. I'm recovering from international travel and should start becoming a bit more engaged on code reviews in the next week

**Review by wesm — COMMENTED** [review]
LGTM, thank you for doing this! 

We should consider raising the profile of this work to Python developers who may be interested in generic high performance IO in both Python and C++, but maybe aren't doing anything with the Arrow columnar format or IPC protocol. This can be useful to a large audience. Would you be interested in writing a brief blog post for arrow.apache.org about this?

**Comment by pitrou** [comment]
By "generic IO" you mean reading and writing non-Arrow data? In that regard I'm not sure we're more interesting than the standard Python IO stack.

By the way, I plan to try and write a blog post about the CSV reader, but before that I'd like to make it a bit more usable.

**Comment by wesm** [comment]
Well, I think the portability aspect of the IO stack is interesting: can be used in C++ only or any language that has bindings. I don't know of a comparable project 

Might be worth waiting to do more blog posts about the IO stack until we have a bit better file system support, eg S3 and Google Cloud Storage, which ought to get done in 2019

**Pull request closed without merging**